project_setup.py

```diff
@@ -0,0 +1,253 @@
+
+
+def replace_file_contents(
+    new_str:str,
+    old_str:str,
+    file_path:str,
+) -> bool:
+    """Replaces instances of `old_str` with `new_str` in file at file_path.
+
+    Args:
+        old_str (str): The string to replace.
+        new_str (str): The string to replace `old_str` with.
+        file_path (str): The path to the file to modify.
+
+    Returns:
+        None
+    """
+    with open(file_path, "r") as f:
+        content = f.read()
+
+    new_content = content.replace(old_str, new_str)
+
+    if new_content != content:
+        with open(file_path, "w") as f:
+            f.write(new_content)
+        return True
+
+    return False
+
+
+def self_destruct(input_func:Callable[[str], str] = input) -> None:
+    """Deletes this file upon confirmation."""
+
+    if input_func("Delete this file? [y/N]: ").strip().lower() == "y":
+        os.remove(__file__)
+
+
+def rtd_project_guess(repo_url:str) -> str:
+    """Guesses the RTD project name from the repo URL."""
+    return repo_url.replace(".git", "").split('/')[-1]
+
+
+def codecov_project_guess(repo_url:str) -> str:
+    """Guesses the codecov project name from the repo URL."""
+    return "/".join(repo_url.replace(".git", "").split('/')[-2:])
+
+
+def run_setup(
+    repo_url:str,
+    package_name:str,
+    dir_path:str = ".",
+    input_func:Callable[[Optional[str]], str] = input,
+    output_func:Callable[[str], None] = print,
+) -> None:
+    # Update pyproject.toml and README.md, which contain the repo URL===========
+    rtd_guess = rtd_project_guess(repo_url)
+    codecov_guess = codecov_project_guess(repo_url)
+
+    replace_args = zip(
+        [repo_url, repo_url, rtd_guess, codecov_guess],
+        [TEMPLATE_REPO_URL, TEMPLATE_REPO_URL,  "ssec-jhu-base-template", "ssec-jhu/base-template"],
+        map(lambda f: os.path.join(dir_path, f), ["pyproject.toml", "README.md", "README.md", "README.md"]),
+        [UPDATED_PYPROJECT_TOML, UPDATED_README_CICD, UPDATED_README_RTD, UPDATED_README_CODECOV]
+    )
+
+    for new_str, old_str, file_path, msg in replace_args:
+        if replace_file_contents(new_str, old_str, file_path):
+            output_func(msg + "✅")
+        else:
+            output_func(msg + "❌")
+    # ==========================================================================
+
+
+    # Update occurences of the package name in the project======================
+    # Sometimes the template package name is referred to using `package_name`
+    # and other times it is referred to by `<package_name>`. First replace the
+    # brackted version becaused replacing the unbracketed version will also
+    # replace the bracketed version, but not vice versa.
+
+    # update the top_level app name dir:
+    os.rename(
+        os.path.join(dir_path, TEMPLATE_PACKAGE_NAME),
+        os.path.join(dir_path, package_name)
+    )
+
+    # get all files with paths in the repo directory
+    dir_file_pairs = itertools.chain.from_iterable(map(
```

Comment:
Here it could be useful also to get info from git, that way it will always leverage gitignores properly (presumably git would be available, would this cause problem on windows and would we be worried then?)

```py
files = subprocess.check_output(['git', 'ls-tree', '-r', '--name-only', 'HEAD']).decode().split('\n')
```

Reply:
Agree using git is probably best. I tried using it on my windows machine and it seemed to work fine. I'll change it.

Reply:
None of this is needed nor advisable, instead, I would use ``git mv``
```python
git.Repo().git.mv("package_name", get_package_name())
```

Besides, not using ``git mv`` is going to nuke the git history, though changing the files with the replace alls may also do this even when using ``git mv``. Might be advisable to commit after each operation to avoid this - additionally this may help the user when things go wrong. I would do ``git mv`` first, commit, and then do all other ops afterward.